### ISSUE TITLE
Question on checking if node is key

### ISSUE THREAD

**Issue by andrei-pavel** [sysrepo/sysrepo#1985]
Dear Michal,
If you could answer this one question, it would save me a lot of time.
I'm trying to find out if a schema node at a certain xpath is key for it's parent list or not. So I wrote this:
```
    bool is_key(std::string const& xpath) {
        S_Data_Node const data_node(session_->get_data(xpath.c_str()));
        Schema_Node_Leaf schema_node_leaf(data_node->schema());
        return schema_node_leaf.is_key() != nullptr;
    }
```
but I get
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Type must be LYS_LEAF
```
And this is because `data_node` is pointing to the root node rather than my nested key candidate leaf whose xpath I gave in the function. This I can figure out by printing `data_node->path()`.

The xpath I give is valid and it points to existing data which is always a leaf, but might not be a key.

So it seems that I cannot retrieve a `Data_Node` pointing directly to a certain xpath? Or a `Schema_Node`? Is there any way I can easily achieve this without iterating with `tree_dfs()` and checking xpath for each node?

Kind regards!

**Comment by michalvasko** [comment]
Hi,
firstly, I apologize for not going into much detail but I already have explained this thoroughly in another issue (and it was quite long, also regarding C++ bindings) and you may be able to understand it anyway. If not, it does not really matter.

So, what you want to do is to get all the data from *sysrepo* you know your XPath may select (it may even be all the data). Then, if you use `find_path()` on these data, you will get a `S_Set` with the actual nodes you expected to get in the first place meaning just the selected ones. And you should be able to make those into leaves and check whether they are keys like you did. Also note that the performance should be much better this way because you will call `get_data()` only once.

Regards,
Michal

**Comment by andrei-pavel** [comment]
Could you take a look at https://github.com/andrei-pavel/sysrepo-example/blob/master/main.cpp?
Relevant functions are `displayXpaths`, `isKey`, `keys`.
I've tried several ways to get data nodes:
```
    // S_Data_Node const data_node(connection_->get_module_info());
    // S_Data_Node const data_node(session_->get_data(MODEL_ALL.c_str()));
    // S_Data_Node const data_node(session_->get_data(MODEL_AND_ROOT_NODE.c_str()));
    S_Data_Node const data_node(session_->get_data(xpath_c_str));
```
I get 0 or 1 elements in the ly_set depending on which one I choose from above. And when I get one element, it has gibberish data, for example the nodetype is unrecognisable.
I've also tried retrieving keys from a list with the `Schema_Node_List::keys()` function. Same thing.
What am I doing wrong? :(
Kind regards

**Comment by michalvasko** [comment]
Hi,
I believe you are over-complicating things. Just make `is_key` parameter an `S_Data_Node` and in the function check that it is a leaf. Then if it is, cast its `node->schema()` to `S_Schema_Node_Leaf` and call `is_key()` on it. But I may be missing something so it is mostly up to you to figure out. I have added an example that does what you are doing, there is nothing more I can do.

Regards,
Michal

**Comment by andrei-pavel** [comment]
Hi,
Thank you for the example.
Managed to make it work with:
```
        S_Context const context(session_->get_context());
        S_Module const module(context->get_module(model_.c_str()));
        S_Schema_Node const schema_node(module->data());
```
so that `find_path()` would return a set with the required node.

I was doing what you say, but unlike the example I needed a way to go directly to the node with the xpath not iterate with `tree_dfs()`. The `tree_dfs()` in my example was only to make the example work with any module and not hardcode an xpath. From the iterated nodes I only got the xpath and I tried to get information solely on that, discarding the node, because that is what I am forced to work with in my real application.

Just so that I have a better understanding, do you happen to know why the other lines in a previous comment above would return empty sets or sets with invalid nodes? They are different types `Data_Node` and `Schema_Node`, but it seems `find_path()` is callable on both.
> ```
>     // S_Data_Node const data_node(connection_->get_module_info());
>     // S_Data_Node const data_node(session_->get_data(MODEL_ALL.c_str()));
>     // S_Data_Node const data_node(session_->get_data(MODEL_AND_ROOT_NODE.c_str()));
>     S_Data_Node const data_node(session_->get_data(xpath_c_str));
> ```

Thank you!

**Comment by michalvasko** [comment]
Hi,
well, then it is hard for me to suggest what to improve if the example is artificially complex to simulate your real scenario. But based on what you said, you only have the data node's xpath and you are trying to learn whether it is the key. You could have said that explicitly.

So, I believe you should simply be able to use `libyang::Context::get_node()`. You use the path you have and it gives you the schema node, which you can easily examine and check whether it is a key. What you are doing is almost the same, except in this case it is nicely packed in a function. Also, while `find_path()` is defined for schema nodes, data nodes, and even the context directly, it does not expect the same path! It is more obvious in the C API because the functions have different prefixes. But, looking for schema nodes generally requires __schema paths__ while for data nodes a __data path__ and you were mixing those. One exception being the function I suggested you use because it is meant precisely for your use-case. What is the difference between these paths and which to use for what function is documented, just follow the links from C++ API to C API.

Regards,
Michal
